File: go/hero_quest/hero_quest_test.go
package hero_quest_test

import (
	"testing"

	"github.com/codecop/Hero-Quest-Refactoring-Kata/hero_quest"
	"github.com/stretchr/testify/assert"
)

var (
	testPlayerName          = "Conan"
	testPlayerHealth        = 100
	testPlayerStrength      = 20
	testPlayerMagic         = 10
	testPlayerCraftingSkill = 10

	testItemName  = "Amulet of Strength"
	testItemKind  = "Strength"
	testItemPower = 10
)

func TestPlayerToString(t *testing.T) {
	result := hero_quest.PlayerToString(testPlayerName, testPlayerHealth, testPlayerStrength, testPlayerMagic, testPlayerCraftingSkill)

	assert.Equal(t, "Conan's Attributes:\nHealth: 100\nStrength: 20\nMagic: 10\nCrafting Skill: 10\n", result)
}

func TestPlayerFallsDown(t *testing.T) {
	testPlayerStrength = 3
	defer func() {
		testPlayerStrength = 20
		testPlayerHealth = 100
	}() // reset

	hero_quest.PlayerFallsDown(&testPlayerHealth, &testPlayerStrength)

	assert.Equal(t, 90, testPlayerHealth)
}

func TestPlayerFallsDownNoDamage(t *testing.T) {
	hero_quest.PlayerFallsDown(&testPlayerHealth, &testPlayerStrength)

	assert.Equal(t, 100, testPlayerHealth)
}

func TestItemToString(t *testing.T) {
	result := hero_quest.ItemToString(testItemName, testItemKind, testItemPower)

	assert.Equal(t, "Item: Amulet of Strength\nKind: Strength\nPower: 10\n", result)
}

func TestItemReduceByUsage(t *testing.T) {
	defer func() {
		testItemPower = 10
	}() // reset
	hero_quest.ItemReduceByUsage(&testItemKind, &testItemPower)

	assert.Equal(t, 5, testItemPower)
}

func TestItemReduceByUsageToJunk(t *testing.T) {
	testItemPower = 1
	defer func() {
		testItemPower = 10
		testItemKind = "Strength"
	}() // reset

	hero_quest.ItemReduceByUsage(&testItemKind, &testItemPower)

	assert := assert.New(t)
	assert.Equal(0, testItemPower)
	assert.Equal("Junk", testItemKind)
}

func TestItemApplyEffectToPlayer(t *testing.T) {
	defer func() {
		testPlayerStrength = 20
	}()
	hero_quest.ItemApplyEffectToPlayer(testItemName, testItemKind, testItemPower, &testPlayerHealth, &testPlayerStrength, &testPlayerMagic)

	assert.Equal(t, 30, testPlayerStrength)
}

func TestItemApplyEffectToPlayerJunk(t *testing.T) {
	hero_quest.ItemApplyEffectToPlayer(testItemName, "Junk", testItemPower, &testPlayerHealth, &testPlayerStrength, &testPlayerMagic)

	assert.Equal(t, 20, testPlayerStrength)
}

func TestItemRepair(t *testing.T) {
	defer func() {
		testItemPower = 10
	}()
	hero_quest.ItemRepair(&testItemPower, testPlayerCraftingSkill)

	assert.Equal(t, 26, testItemPower)
}

File: java/src/test/java/codingdojo/HeroQuestTest.java
package codingdojo;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class HeroQuestTest {

    private String playerName;
    private int playerHealth;
    private int playerStrength;
    private int playerMagic;
    private int playerCraftingSkill;
    private String itemName;
    private String itemKind;
    private int itemPower;

    @BeforeEach
    void SetUp() {
        HeroQuest.output.setLength(0);
        playerName = "Conan";
        playerHealth = 100;
        playerStrength = 20;
        playerMagic = 10;
        playerCraftingSkill = 10;
        itemName = "Amulet of Strength";
        itemKind = "Strength";
        itemPower = 10;
    }

    @Test
    void playerToString() {
        var result = HeroQuest.playerToString(playerName,
                playerHealth, playerStrength, playerMagic, playerCraftingSkill);

        var expected = "Conan's Attributes:\nHealth: 100\nStrength: 20\nMagic: " +
                "10\nCrafting " +
                "Skill: 10\n";

        assertEquals(expected, result);
    }

    @Test
    void playerFallsDown() {
        playerStrength = 3;
        playerHealth = HeroQuest.playerFallsDown(playerStrength, playerHealth);
        assertEquals(90, playerHealth);
    }

    @Test
    void playerFallsDownNoDamage() {
        playerHealth = HeroQuest.playerFallsDown(playerStrength, playerHealth);
        assertEquals(100, playerHealth);
    }

    @Test
    void itemToString() {
        var result = HeroQuest.itemToString(itemName, itemKind, itemPower);
        var expected = "Item: Amulet of Strength\nKind: Strength\nPower: 10\n";
        assertEquals(expected, result);
    }

    @Test
    void itemReduceByUsage() {
        HeroQuest.ItemUsageResult result = HeroQuest.itemReduceByUsage(itemKind, itemPower);
        itemPower = result.itemPower;
        assertEquals(5, itemPower);
    }

    @Test
    void itemReduceByUsageToJunk() {
        itemPower = 1;
        HeroQuest.ItemUsageResult result = HeroQuest.itemReduceByUsage(itemKind, itemPower);
        itemPower = result.itemPower;
        itemKind = result.itemKind;
        assertEquals(0, itemPower);
        assertEquals("Junk", itemKind);
    }

    @Test
    void itemApplyEffectToPlayer() {
        HeroQuest.ItemEffectResult result = HeroQuest.itemApplyEffectToPlayer(itemName, itemKind, itemPower, playerHealth, playerStrength, playerMagic);
        playerStrength = result.playerStrength;
        assertEquals(30, playerStrength);
    }

    @Test
    void itemApplyEffectToPlayerJunk() {
        itemKind = "Junk";
        HeroQuest.ItemEffectResult result = HeroQuest.itemApplyEffectToPlayer(itemName, itemKind, itemPower, playerHealth, playerStrength, playerMagic);
        playerStrength = result.playerStrength;
        assertEquals(20, playerStrength);
    }

    @Test
    void itemRepair() {
        itemPower = HeroQuest.itemRepair(playerCraftingSkill, itemPower);
        assertEquals(26, itemPower);
    }
}
